cpp/include/ray/api/ray_register.h

```diff
@@ -0,0 +1,201 @@
+          break;
+        }
+
+        it->second(data, size, result);
+        if (result.size() >= MAX_BUF_LEN) {
+          result = PackArgsStr(
+              ErrorCode::FAIL,
+              "the response result is out of range: more than 10M " + func_name);
+        }
+      } catch (const std::exception &ex) {
+        result = PackArgsStr(ErrorCode::FAIL, ex.what());
+      }
+    } while (0);
+
+    return result;
+  }
+
+ private:
+  Router() = default;
+  ~Router() = default;
+  Router(const Router &) = delete;
+  Router(Router &&) = delete;
+
+  template <typename F, size_t... I, typename Arg, typename... Args>
+  static typename std::result_of<F(Args...)>::type CallHelper(
+      const F &f, const absl::index_sequence<I...> &, std::tuple<Arg, Args...> tup) {
+    return f(std::move(std::get<I + 1>(tup))...);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      !std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    auto r = CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK, "ok", r);
+  }
+
+  template <typename Function>
+  struct invoker {
+    static inline void Apply(const Function &func, const char *data, size_t size,
+                             std::string &result) {
+      using args_tuple = AddType_t<std::string, boost::callable_traits::args_t<Function>>;
+
+      try {
+        auto tp = Serializer::Deserialize<args_tuple>(data, size);
+        Router::Call(func, result, std::move(tp));
+      } catch (msgpack::type_error &e) {
+        result =
+            PackArgsStr(ErrorCode::FAIL, std::string("invalid arguments: ") + e.what());
+      } catch (const std::exception &e) {
+        result = PackArgsStr(ErrorCode::FAIL,
+                             std::string("function execute exception: ") + e.what());
+      } catch (...) {
+        result = PackArgsStr(ErrorCode::FAIL, "unknown exception");
+      }
+    }
+  };
+
+  template <typename Function>
+  void RegisterNonmemberFunc(std::string const &name, Function f) {
+    this->map_invokers_[name] = {std::bind(&invoker<Function>::Apply, std::move(f),
+                                           std::placeholders::_1, std::placeholders::_2,
+                                           std::placeholders::_3)};
+  }
+
+  template <typename T>
+  static std::string PackArgsStr(int error_code, std::string error_msg, T result) {
```

Comment:
It seems that this function is only used to serialize return values. It has nothing to do with arguments. Better to rename it to `PackReturnValue`. if you want to keep "Str", please add "To" before it, like "PackxxxToStr".

Reply:
Good